src/Refinery/Tactic.hs

```diff
@@ -62,27 +80,67 @@ infixr 3 <%>
 (<%>) :: TacticT jdg ext err s m a -> TacticT jdg ext err s m a -> TacticT jdg ext err s m a
 t1 <%> t2 = tactic $ \j -> Interleave (proofState t1 j) (proofState t2 j)
 
--- | @commit t1 t2@ will run @t1@, and then only run @t2@ if @t1@ failed to produce any extracts.
-commit :: TacticT jdg ext err s m a -> TacticT jdg ext err s m a -> TacticT jdg ext err s m a
-commit t1 t2 = tactic $ \j -> Commit (proofState t1 j) (proofState t2 j)
-
 -- | Tries to run a tactic, backtracking on failure
 try :: (Monad m) => TacticT jdg ext err s m () -> TacticT jdg ext err s m ()
 try t = t <|> pure ()
 
--- | Runs a tactic repeatedly until it fails
+-- | @commit t1 t2@ will run @t1@, and then run @t2@ only if @t1@ failed to produce any successes.
+--
+-- NOTE: @commit@ does have some sneaky semantics that you have to be aware of. Specifically, it interacts a bit
+-- surprisingly with '>>='. Namely, the following algebraic law holds
+-- @
+--     commit t1 t2 >>= f = commit (t1 >>= f) (t2 >>= f)
+-- @
+commit :: TacticT jdg ext err s m a -> TacticT jdg ext err s m a -> TacticT jdg ext err s m a
+commit t1 t2 = tactic $ \j -> Commit (proofState t1 j) (proofState t2 j)
+
+-- | Runs a tactic 0 or more times until it fails.
+-- Note that 'many_' is almost always the right choice over 'many'.
+-- Due to the semantics of 'Alternative', 'many' will create a bunch
+-- of partially solved goals along the way, one for each iteration.
```

Comment:
I think you could get away with an `Applicative` constraint instead if you used `*>`

Reply:
Sadly not, due to that `Effect` ctor